Combo set of async and other changes

qkoziol opened the pull request:
Draft PR for all code to merge to develop, includes async, dynamic optional VOL operation registration, and H5S_BLOCK code.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Update API tracing for new H5VL_request_status_t typedef

qkoziol committed:
Finish converting internal event set operations to use list iterator callbacks, instead of directly accessing the list structure

qkoziol committed:
Add H5VL_REQUEST_GET_ERR_STACK operation to request subclass, for retrieving a copy of the error stack for a failed asynchronous operation

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Remove 'canceled' event status from Java constants

qkoziol committed:
Be safer about releasing resources when inserting a newly opened/created object or file into an event set

qkoziol committed:
Remove H5EStest, add H5ES_WAIT_NONE for 0 timeout, and revise parameters to H5ESwait, to make it more "aggregate".

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Remove H5ES_STATUS_CANCELED from Java wrappers also

qkoziol committed:
Apply patch for dynamically registering optional VOL operations

Commit:
(a) Add async APIs for H5O module as listed in jira issue ID-283.
(b) Remove verification of name parameter in async related routines for H55A and H5L modules
    because it is checked in H5VL_setup* routine.
(c) Modify h5dump expected output due to the async changes.

Commit:
Corrections based on PR feedback.

Commit:
Further changes to make based on PR feedback.

vchoi-hdfgroup committed:
Merge pull request #98 from vchoi-hdfgroup/async

(a) Add async APIs for H5O module as listed in jira issue ID-283.

qkoziol committed:
Remove H5Dwait & H5Fwait (moved to the async connector).  Added H5atclose
routine.  Updated 'optional op' operations.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Fix missed merge marker, and reformatted line

qkoziol committed:
Update API tracing infrastructure for H5atclose callback

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Clean up some warnings

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Normalize against develop branch

qkoziol committed:
Correct level of indirection

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Add doxygen info for H5is_library_terminating and regression tests for it and H5atclose

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Relocate prototype (and doxygen info) for H5Aclose

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Align w/changes on develop

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Move group package initialization code to H5Gint.c, and update tracing macros

qkoziol committed:
Change non-static function declarations to be static

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Correct GCC diagnostic macro

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

jhendersonHDF committed:
Ensure that H5TSpublic.h header gets installed (#129)

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Finish moving API routines that invoke VOL framework to main source files.

qkoziol committed:
Fix position of H5Fmount and H5Funmount

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Add 'wrapper' versions of async calls, to allow language wrappers and layers on top of HDF5 to pass in their application information.

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Add wrappers for dynamically registered optional operations

qkoziol committed:
Fix typo

qkoziol committed:
Update doxygen comment for H5atclose with additional detail.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Add H5VL\*_vararg versions of H5VL routines that use va_list parameters

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Implement and test H5S_BLOCK

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into packed_mem_buffer

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Switch H5Aexists\*_async and H5Lexists\*_async to use flag to return status, instead of return value.  Make the corresponding changes through most of the v1 and v2 B-tree code.  Clean up warnings in H5public.h and cmpd_dtransform.c.

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Add H5Iregister_future routine and tests.

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Correct return value for H5Lexists_async

qkoziol committed:
Add H5_DLL macro to public H5ES API routines

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Update supported -> flags parameter for introspect_query callback

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Remove my email address.  Update passthrough VOL connector ID.

qkoziol committed:
Fix comment for post_open_api_common

qkoziol committed:
Remove unused non-blocking VOL connector

Commit:
Minor cleanup in async branch in preparation for merge to develop

Commit:
Update CMake and the Autotools to use the new pass-through VOL ID

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Finish another iteration on public H5ES routines, along with running the code reformatter

qkoziol committed:
Merge remote-tracking branch 'origin/packed_mem_buffer' into async_vol_register_optional

qkoziol committed:
Another round of reformatting

derobins committed:
Fix for SWMR daily test failures (#160)

The H5I_register_using_existing_id() call did not initialize the future
ID callbacks, causing the library to segfault when it tried to
resolve those function pointers.

derobins committed:
Merge pull request #158 from derobins/async

Minor cleanup in async branch in preparation for merge to develop

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

bmribler committed:
Added selective async APIs (#150)

* Added selective async APIs

Description:
    Added the following APIs:
        H5Ropen_attr_async
        H5Ropen_object_async
        H5Ropen_region_async

        H5Mcreate_async
        H5Mopen_async
        H5Mput_async
        H5Mget_async
        H5Mclose_async

        H5Tcommit_async
        H5Topen_async
        H5Tcopy_async
        H5Tclose_async
    - Updated an expected output file to include a new internal function
      in the error stack for the failure case.

* Updated async APIs per reviews, including removing async version of
H5Tcopy.

* Removed statements that were added by mistake in the previous commit.

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Fix compile issues in H5M and warnings elsewhere

qkoziol committed:
Reformat code

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

derobins committed:
Brings VOL_LIST changes from develop. (#163)

derobins committed:
Merge branch 'develop' into async

qkoziol committed:
Merge remote-tracking branch 'origin/async' into async_vol_register_optional

qkoziol commented:
This is the "combo" version of the async changes, plus the dynamic optional VOL operations, etc.   It's ready to merge after the base async branch is merged to develop.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol commented:
I've re-synced this branch with develop, after the async branch was merged in.  This should be ready to merge also now, after develop is tagged for the DAOS team.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Remove H5Dwait and H5Fwait calls, which were incorrectly brought back in

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Tiny cleanup of H5Lcreate_hard_async

gnuoyd commented on src/H5ESpublic.h:
```diff
@@ -42,11 +42,26 @@
+    uint64_t op_ins_count;  /* Counter of operation's insertion into event set */
+    uint64_t op_ins_ts;     /* Timestamp for when the operation was inserted into the event set */
+    uint64_t op_exec_ts;    /* Timestamp for when the operation began execution */
+    uint64_t op_exec_time;  /* Execution time for operation */
```
Units?

gnuoyd commented on src/H5ESint.c:
```diff
@@ -232,6 +249,78 @@ H5ES__create(void)
+ *	        Friday, December 11, 2020
+ *
+ *-------------------------------------------------------------------------
+ */
```
You have spent 10 lines "documenting" H5ES__insert without telling me the meaning of a single argument. I can infer the return values from the return type, `herr_t`.

markcmiller86 replied:
@epourmal for some reason, in my mind's eye, I had the impression the new docs-like-code workflow THG started in the PSIP work and has been working towards would result in being able to document functions more like this...

```
static herr_t
 H5ES__insert(
    H5ES_t *es,   /<** the event set object identifier **/
    H5VL_t *connector, /**< the VOL connector identifier **/
    void *request_token, /**< the request token, 
    const char *app_file, /**< the name of the originating application file of the request **/
    const char *app_func, /**< the name of the originating application function of the request **/
    unsigned app_line, /**< the line of application source code of the originating request **/
    const char *caller, /**< the name of the calling function **/
    const char *api_args) /**< the calling function's arguments, separated by commas **/
```

and have that get slurrped into the browseable documentation using whatever automation doc tool workflow you use. Am I totally misunderstanding things?

Also, @gnuoyd, apologies for inserting myself into your conversation here but I am curious...I have read absolutely nothing and have discussed absolutely nothing with anyone about this API and still believe I know what (most of) the arguments to this function are (supposed to be) because of how they are named. I don't know what the difference is between `app_` args and `caller` and `api_args` and I would want further clarification on that. However, I wonder how much more useful the additional documentation (or a version of it that you would prefer) really leads to deeper/better understanding.

gnuoyd replied:
@markcmiller86, I find the argument names meaningful, too. I object to the 10-line header comment precisely because it takes up so much space while telling us little more about the program than the code itself does. An unadorned imperative sentence will suffice, `/* Insert a request token into an event set. */`.

The current comment is a waste of display space that could hold program meaning, instead.

qkoziol replied:
> @epourmal for some reason, in my mind's eye, I had the impression the new docs-like-code workflow THG started in the PSIP work and has been working towards would result in being able to document functions more like this...
> 
> ```
> static herr_t
>  H5ES__insert(
>     H5ES_t *es,   /<** the event set object identifier **/
>     H5VL_t *connector, /**< the VOL connector identifier **/
>     void *request_token, /**< the request token, 
>     const char *app_file, /**< the name of the originating application file of the request **/
>     const char *app_func, /**< the name of the originating application function of the request **/
>     unsigned app_line, /**< the line of application source code of the originating request **/
>     const char *caller, /**< the name of the calling function **/
>     const char *api_args) /**< the calling function's arguments, separated by commas **/
> ```
> 
> and have that get slurped into the browseable documentation using whatever automation doc tool workflow you use. Am I totally misunderstanding things?

Currently, the Doxygen comments are only in the HDF5 public header files.   Adding them to the parameters for internal routines may be useful if the coding standard was updated to push in that direction.  (Although it would be best done incrementally, in the Boy Scout "Leave the code you change in better shape than you found it" way, since it would be a _huge_ job overall :-)

gnuoyd commented on src/H5ESint.c:
```diff
@@ -232,6 +249,78 @@ H5ES__create(void)
+        HGOTO_ERROR(H5E_EVENTSET, H5E_CANTCREATE, FAIL, "can't create event object")
+
+    /* Copy the app source information */
+    if (NULL == (ev->op_info.app_file_name = H5MM_xstrdup(app_file)))
```
I was concerned about unnecessary overheads when I saw that you were passing the source filename, function name, and line number. Now I see that you are strdup'ing some of the information. This is an awful lot of overhead for an application to absorb when it's not being debugged. Let's talk on the mailing list about how to lower the costs and put this under some kind of runtime switch.

gnuoyd commented on src/H5ES.c:
```diff
@@ -135,6 +184,41 @@ H5ESget_count(hid_t es_id, size_t *count /*out*/)
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5ESget_time_estimate(hid_t es_id, uint64_t *time_est /*out*/)
```
I don't think we should add `get_time_estimate` to the library API until there are some example implementations that provide it and some evidence that any application actually needs it.

derobins replied:
I am also not a fan of this API call. I think it's going to be nigh impossible to make it useful.

qkoziol replied:
It's up to the VOL connector to figure it out, and it's a "quality of implementation" aspect of an asynchronous VOL connector.   Any connector that doesn't want to try to estimate operations can always return the "unknown estimate" value (UINT64_MAX) for a request and the event set will stop iterating over the requests in the event set and return the "unknown" value to the application.

I would have used the new "dynamic optional VOL operations" feature for it, but it requires the addition of a new VOL operation value for requests and there's no way to have the event set "know" about a dynamically registered operation.

LBNL is currently improving the async VOL connector to track the I/O statistics and return more meaningful time estimates, and we will work with applications to take advantage of this capability to balance their compute vs. I/O.  Having a time estimate for a set of I/O operations is a _big_ enabler for pushing applications to take advantage of async I/O - it's something they've never had the ability to predict before.   When an application retrieves the time estimate and it's a lot larger or smaller than their predicted compute time, they can adjust one or the other to better overlap and take more advantage of the system resources.

markcmiller86 replied:
Just spit-balling here but might there be a way here or in the future for a record of time an `unknown estimate` operation wound up taking and a place to store/hash that value so that if the situation is encountered again, instead of `unknown estimate` it can return that value based on previous experience? Thats likely well beyon what is necessary to get this initial implementation into the main lib but wanted to toss out the idea to possibly inform thinking.

qkoziol replied:
Good point - still up the VOL connector, but could be useful for that code.

gnuoyd commented on src/H5VLcallback.c:
```diff
@@ -3526,6 +4171,50 @@ H5VLfile_get(void *obj, hid_t connector_id, H5VL_file_get_t get_type, hid_t dxpl
+ */
+herr_t
+H5VLfile_get_vararg(void *obj, hid_t connector_id, H5VL_file_get_t get_type, hid_t dxpl_id,
+                    void **req /*out*/, ...)
```
Just reading this is giving me tendonitis. :-)

It is not necessary to be so repetitive. Let's talk.

gnuoyd reviewed (changes requested):
Is there some design document for the optional operations?

There is too much going on in this one PR and it should be broken down.

qkoziol commented:
> Is there some design document for the optional operations?
> 
> There is too much going on in this one PR and it should be broken down.

I completely understand, it was not my original intention to submit it in this manner either, but that was what we agreed to do last Friday.  I will continue to produce documentation, tests, code updates, etc. after it's merged in.   Please release your "change request" hold and add items to the mind map here for prioritization and discussion:  https://coggle.it/diagram/X9e7HxlvmVpHJ7Rb/t/ecp-deliverables

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Run source formatter

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Allow for canceled operation in wait_cb

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Attempt to fix switch on string value

soumagne commented:
I see a lot of formatting changes in that PR, can someone explain to me how this can be possible? (@byrnHDF) I thought that once we moved to clang-format everything was done automatically and checked whenever a new PR is made?

byrnHDF commented:
That should be the case - we will need to investigate why.

Allen

________________________________________
From: Jerome Soumagne <notifications@github.com>
Sent: Thursday, January 7, 2021 11:53 AM
To: HDFGroup/hdf5
Cc: Allen Byrne; Mention
Subject: Re: [HDFGroup/hdf5] Combo set of async and other changes (#161)

I see a lot of formatting changes in that PR, can someone explain to me how this can be possible? (@byrnHDF<https://github.com/byrnHDF>) I thought that once we moved to clang-format everything was done automatically and checked whenever a new PR is made?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5/pull/161#issuecomment-756275879>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AL77KBTDDURLASCU2TM6LFTSYXYIZANCNFSM4UXHSFKQ>.

byrnHDF commented:
> That should be the case - we will need to investigate why. Allen
I see 
`clang-format on`
comments that would not be in format checks until the next line.
I saw some other comments with "//" that might need special handling, maybe?

byrnHDF commented:
As for the others, they must have been checked in without running format?

qkoziol commented:
Hmm, I’ll re-run the formatter and push any change to the PR.

qkoziol committed:
Re-run source formatter

soumagne commented:
If the current develop needs the formatter to be run, it should be done separately from that PR.

qkoziol commented:
> If the current develop needs the formatter to be run, it should be done separately from that PR.

Yep, submitting one for develop now.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol commented:
In addition to the mind map, I’m working on a support document for this PR, here:   https://docs.google.com/document/d/1xNOq3MfLp7YhKlmxOKoZzag03q0YoHONNiz9svQ60Iw/edit?usp=sharing

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

epourmal commented on src/H5ESpublic.h:
```diff
@@ -120,13 +120,18 @@ extern "C" {
 H5_DLL herr_t H5ESwait(hid_t es_id, uint64_t timeout, size_t *num_in_progress, hbool_t *err_occurred);
+H5_DLL herr_t H5EScancel(hid_t es_id, size_t *num_not_canceled, hbool_t *err_occurred);
 H5_DLL herr_t H5ESget_count(hid_t es_id, size_t *count);
+H5_DLL herr_t H5ESget_time_estimate(hid_t es_id, uint64_t *time_estimate);
```
This function still needs more discussions based on the review on 1/8/21

qkoziol commented:
> In addition to the mind map, I’m working on a support document for this PR, here: https://docs.google.com/document/d/1xNOq3MfLp7YhKlmxOKoZzag03q0YoHONNiz9svQ60Iw/edit?usp=sharing

I’ve finished updating this support document to cover all of the changes in the PR.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

markcmiller86 commented on test/vol.c:
```diff
@@ -162,6 +295,85 @@ static const H5VL_class_t fake_vol_g = {
     NULL /* optional     */
 };
 
+/*-------------------------------------------------------------------------
```
Does this test code demonstrate/exercise the API as it might be used by in a real use case or is it really just unit-checking the API? If the latter, I would want to see a real and well-documented end-user example which exercises key (if not all) modalities of use (or separate tests for separate modalities) also with this code base to understand how the user-facing API is intended to be used.

qkoziol replied:
This is a unit test and is only checking the API for correctness.   There’s a separate VOL connector testing suite under development, here:   https://github.com/HDFGroup/vol-tests

markcmiller86 commented:
I happened to be browsing around here and just had a few questions. Not urgent to answer so don't let any of my questions get in the way of the real work of completing this PR...

- GitHub provides a pretty useful wiki which supports a wide variety of document formats include markdown but also sphinx, wikimedia, etc. I wonder if it makes sense to handle supporting documentat like the google doc referenced from this issue in that form instead? Maybe non-essential but I find it helps to keep all the docs in one place
- This PR seems like it is mixing a few different things. Any reason it cannot be decomposed into separate PRs each for specific functionality needed? Turns out [I've been writing](https://github.com/betterscientificsoftware/bssw.io/blob/cc-mcm86-10dec20-large-prs/Articles/ShortArticles/PullRequestSizeConsiderations-wikized.md) on just this issue recently.
- The google doc refers to a use case of sorts. Is it possible to have that end-user use case as a test/example code? With that, as API changes, it too would get updated. In addition, it could be documented such that it kinda sorta serves as the key umbrella artifact to facilitate dialog about the change and its specific details and why they are needed -- or maybe alternatives to consider.
- Are 14 reviewers really needed to get this through? That seems like huge overkill and a big use of resource/time. On VisIt, (2.8 million lines of C++ code with some Python) we handle PRs that are probably 10-20x this size with one or two reviewers. We also try to complete PR reviews in < 2-3 days. I don't think I've seen any project that involves 14 reviewers for a PR. Just curious what the rationale is.
- Just curious but in your testing (which is way cool BTW!!), do you have an idea on the coverage yet? Would be a nice warm fuzzy if the tests here for this new functionality were known to cover most of the code.
- If one wanted to submit doc updates with these code changes (e.g. docs-like-code), is that possible here?

byrnHDF commented on java/src/hdf/hdf5lib/HDFArray.java:
```diff
@@ -169,10 +169,6 @@ else if (ArrayDescriptor.className
                                 .equals("java.lang.Integer")) {
                             therow = IntegerToByte((Integer[]) _theArray);
                         }
-                        else if (ArrayDescriptor.className
```
Why was this removed?

qkoziol replied:
The changes in this file were to correct Java compile failures on one of the THG machines (sorry, can’t remember which one, maybe @lrknox remembers), which were failing with a complaint about a switch using a string variable.

byrnHDF replied:
I think this file and the changes need to be extracted to a separate issue for further investigation.

byrnHDF replied:
Created PR just for this file

byrnHDF commented on java/src/hdf/hdf5lib/HDFArray.java:
```diff
@@ -522,36 +518,34 @@ else if (ArrayDescriptor.className.equals("java.lang.Long")) {
             flattenedArray = (Object) _barray;
             break;
         case 'L':
-            switch (ArrayDescriptor.className) {
```
Is this file maybe not up-to-date?

byrnHDF replied:
I can understand the changes for this switch, again this should be extracted to a separate issue.

byrnHDF commented on test/Makefile.am:
```diff
@@ -176,7 +176,7 @@ CHECK_CLEANFILES+=accum.h5 cmpd_dset.h5 compact_dataset.h5 dataset.h5 dset_offse
     earray_hdr_fd.h5 farray_hdr_fd.h5 bt2_hdr_fd.h5 \
     storage_size.h5 dls_01_strings.h5 power2up.h5 version_bounds.h5 \
-    alloc_0sized.h5 \
+    alloc_0sized.h5 h5s_block.h5 \
```
Should also be added to CMake file.

qkoziol replied:
Yes, will add, thanks!

qkoziol committed:
Add H5S_BLOCK testfile to CMake clean target

qkoziol commented:
> I happened to be browsing around here and just had a few questions. Not urgent to answer so don't let any of my questions get in the way of the real work of completing this PR...
> 
> * GitHub provides a pretty useful wiki which supports a wide variety of document formats include markdown but also sphinx, wikimedia, etc. I wonder if it makes sense to handle supporting documentat like the google doc referenced from this issue in that form instead? Maybe non-essential but I find it helps to keep all the docs in one place

Sounds like a good idea, I’ll try the wiki for future situations like this.

> * This PR seems like it is mixing a few different things. Any reason it cannot be decomposed into separate PRs each for specific functionality needed? Turns out [I've been writing](https://github.com/betterscientificsoftware/bssw.io/blob/cc-mcm86-10dec20-large-prs/Articles/ShortArticles/PullRequestSizeConsiderations-wikized.md) on just this issue recently.

I agree, but there were complex issues around this PR, outside of the code proper, and it worked out that all of the changes should be put into one PR.

> * The google doc refers to a use case of sorts. Is it possible to have that end-user use case as a test/example code? With that, as API changes, it too would get updated. In addition, it could be documented such that it kinda sorta serves as the key umbrella artifact to facilitate dialog about the change and its specific details and why they are needed -- or maybe alternatives to consider.

Yes, I’ll see if I can work that into the VOL testing code (in https://github.com/HDFGroup/vol-tests)

> * Are 14 reviewers really needed to get this through? That seems like huge overkill and a big use of resource/time. On VisIt, (2.8 million lines of C++ code with some Python) we handle PRs that are probably 10-20x this size with one or two reviewers. We also try to complete PR reviews in < 2-3 days. I don't think I've seen any project that involves 14 reviewers for a PR. Just curious what the rationale is.

Agree.   Some of the reviewers are auto-assigned, though.

> * Just curious but in your testing (which is way cool BTW!!), do you have an idea on the coverage yet? Would be a nice warm fuzzy if the tests here for this new functionality were known to cover most of the code.

There’s no code coverage testing on HDF5 currently.   I agree with you, that would be a really good addition. :-)

> * If one wanted to submit doc updates with these code changes (e.g. docs-like-code), is that possible here?

To some degree - the Doxygen comments in the public headers could be updated in a PR on the HDF5 repo, but much of the other HDF5 documentation is in other repos, and I don’t think the main hdf5doc repo has been moved to github (yet?), it’s still on the THG bitbucket:  https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5doc/browse

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Add H5Pset_vol_async API routine and 'get_cap_flags' VOL introspection callback

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Clean up warnings

qkoziol committed:
Add H5P(set\|get)_vol_implicit_async API routines to allow \/ disallow implicit asynchronous operations (default is disallowed)

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Run formatting script

qkoziol committed:
Remove H5VL_REQUEST_WAIT\*

qkoziol committed:
Warning cleanup

qkoziol committed:
Eliminate strdup()s on statically allocated strings

qkoziol committed:
Warning cleanup

qkoziol committed:
Split H5VLrestore_lib_state into H5VLstart_lib_state and H5VLrestore_lib_state, and rename H5VLreset_lib_state to H5VLfinish_lib_state.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Duplicate strings when building err_info to return to applicatin

qkoziol committed:
Move connector author routines into seperate header files, all included in the new hdf5dev.h header

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Run bin/trace to add TRACE macros

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Allow H5ES_NONE as a valid, but no-op, parameter to all H5ES API routines that accept an event set ID

qkoziol committed:
Clean up formatting

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Remove H5Pset/get_vol_implicit_async

epourmal commented:
Hi Quincey,

I preparation for the merge, could you please add descriptions of the changes to the RELEASE.txt file? Thank you!

epourmal commented:
Would it be possible to indicate which changes should go into 1_12, 1_10 and (maybe?) 1_8 branches? Thank you!

qkoziol commented:
> Hi Quincey,
> 
> I preparation for the merge, could you please add descriptions of the changes to the RELEASE.txt file? Thank you!

Yes, I will definitely do this.

qkoziol commented:
> Would it be possible to indicate which changes should go into 1_12, 1_10 and (maybe?) 1_8 branches? Thank you!

If I have some time, I will go back to annotate merges to prior release branches.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Clean up warning

qkoziol committed:
Remove H5Pget_vol_async and replace with more generic H5Pget_vol_cap_flags

qkoziol committed:
Clean up warnings

qkoziol committed:
Add H5ESfree_err_info convenience routine

qkoziol committed:
Fix typo

qkoziol committed:
Correct matching for cached VOL plugins

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Add developer header file

qkoziol committed:
Update for C99 compatibility

qkoziol committed:
Add missing trace macro

qkoziol committed:
Stop clang-format from messing with the trace macros.  Don't set up VOL wrappers for 'infrastructure' objects like requests and blobs

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Fix warning about formatting a directory

qkoziol committed:
Clean up formatting for H5E_BEGIN_TRY / H5E_END_TRY

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Reduce scope of H5ES__close

qkoziol committed:
Enable CMake checks for various types on MacOS

qkoziol committed:
Clean up properly when H5CX_retrieve_state() fails.  Also clean up many compiler warnings.

github-actions[bot] committed:
Committing clang-format changes

byrnHDF reviewed (approved):
The CMake, java and tools look fine.

qkoziol committed:
Merge from develop

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Fix mis-placed assert

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

jhendersonHDF commented on src/H5ESevent.c:
```diff
@@ -185,6 +185,14 @@ H5ES__event_completed(H5ES_event_t *ev, H5ES_event_list_t *el)
+                  "%s: releasing event for '%s' (count: %llu, timestamp: %llu), with args '%s', in app source
+       " "file '%s', function '%s', and line %u\n", FUNC, ev->op_info.api_name, (unsigned long
+       long)ev->op_info.op_ins_count, (unsigned long long)ev->op_info.op_ins_ts, ev->op_info.api_args,
+       ev->op_info.app_file_name, ev->op_info.app_func_name, ev->op_info.app_line_num);
```
Would it be possible to loop this into HDF5's debugging capabilities? Not much reason to leave a commented out fprintf statement in the library.

qkoziol replied:
There's already similar info available in the error info for failed async tasks, I'll take this debugging printf out.

jhendersonHDF commented on src/H5TSdevelop.h:
```diff
@@ -11,14 +11,15 @@
-/* Public headers needed by this file */
-#include "H5public.h" /* Generic Functions                    */
+/* Include package's public header */
+/* #include "H5TSpublic.h" */
```
Any need to keep this line around?

qkoziol replied:
There are similar placeholders in the library, so that's why I had left it in.   But, there's plenty of examples of how / where to include a public header file for an interface, so I'll take it out.

jhendersonHDF commented on src/H5TSprivate.h:
```diff
@@ -25,8 +25,9 @@
-/* Public headers needed by this file */
-#include "H5TSpublic.h" /* Public API prototypes */
+/* Include package's public headers */
+/* #include "H5TSpublic.h" */
```
Same as in H5TSdevelop.h - any need to keep this line?

qkoziol replied:
Ditto - will take out

jhendersonHDF commented on src/H5VLcallback.c:
```diff
@@ -1437,6 +1486,50 @@ H5VLattr_get(void *obj, hid_t connector_id, H5VL_attr_get_t get_type, hid_t dxpl
+ */
+herr_t
+H5VLattr_get_vararg(void *obj, hid_t connector_id, H5VL_attr_get_t get_type, hid_t dxpl_id,
+                    void **req /*out*/, ...)
```
I may very well have missed the discussions surrounding adding var_args versions of the VOL callbacks, but did we end up coming to a conclusion on these? While it looks like it did help clean up the passthrough connector a bit, it seems very inconvenient to have to add (and maintain) separate versions of each VOL call. Not to mention that it seems like a connector author may have problems determining when one is needed vs. the other.

qkoziol replied:
Yes, we discussed adding them, a while ago, so they are in.   Having two versions is overkill to me also, but the va_list version was "legacy" at the time I added the vararg version.   If we're happy with it enough to justify [another] breaking change to the VOL connector callbacks, I can remove the va_list version.

jhendersonHDF replied:
Ok, I think we should definitely discuss this with everyone then to see what the consensus is. IMO, it's still a good time to finalize any breaking changes, and, again IMO, having two versions will probably just end up being a nightmare for both maintenance and cognitive overhead in the long run.

qkoziol replied:
> Ok, I think we should definitely discuss this with everyone then to see what the consensus is. IMO, it's still a good time to finalize any breaking changes, and, again IMO, having two versions will probably just end up being a nightmare for both maintenance and cognitive overhead in the long run.

Agree - I'll add it to the agenda for next Monday's technical meeting

fortnern replied:
Elena is open to changing the existing functions to use varargs

jhendersonHDF commented on src/H5A.c:
```diff
@@ -1057,15 +1057,15 @@ H5Aread(hid_t attr_id, hid_t dtype_id, void *buf /*out*/)
     herr_t         ret_value = SUCCEED;         /* Return value */
 
     FUNC_ENTER_API(FAIL)
-    H5TRACE7("e", "*s*sIuii*xi", app_file, app_func, app_line, attr_id, dtype_id, buf, es_id);
```
Looks like the *x -> x in this line and the change below might be unintentional?

qkoziol replied:
No, that's intentional and the result of the bin/trace script updates that parse the parameters better.

jhendersonHDF replied:
This is probably just from my ignorance of bin/trace in general, but the only changes to bin/trace I see are adding cases for H5ES_event_complete_func_t and H5ES_event_insert_func_t. How does that affect needing the asterisk for the void * buf parameter here?

qkoziol replied:
> This is probably just from my ignorance of bin/trace in general, but the only changes to bin/trace I see are adding cases for H5ES_event_complete_func_t and H5ES_event_insert_func_t. How does that affect needing the asterisk for the void * buf parameter here?

Adding the "/*out*/" comment to the parameters indicates to the bin/trace script that the buffer contents are not valid at function entry and are therefore an 'x' (i.e. totally opaque to tracing) and not a '*x' (which indicates a pointer to a buffer of bytes).

jhendersonHDF commented on src/H5D.c:
```diff
@@ -1032,13 +1036,13 @@ H5Dread_async(const char *app_file, const char *app_func, unsigned app_line, hid
     herr_t         ret_value = SUCCEED; /* Return value */
 
     FUNC_ENTER_API(FAIL)
-    H5TRACE5("e", "ii*h*Iu*x", dset_id, dxpl_id, offset, filters, buf);
```
The same as for H5Aread_async - is the argument change here intentional?

qkoziol replied:
Yes

qkoziol committed:
Remove commented-out code

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

jhendersonHDF commented on test/vol.c:
```diff
@@ -1176,6 +1538,510 @@ test_basic_datatype_operation(void)
+    if (conn_env_str) {
+        if (NULL == (conn_env_str = HDstrdup(conn_env_str)))
+            TEST_ERROR
+        if (HDunsetenv("HDF5_VOL_CONNECTOR") < 0)
```
Seems Windows is unhappy about this, since setenv/unsetenv aren't standard C. @derobins Is there a Windows equivalent for these? They're POSIX 2001, so I figure there probably are.

qkoziol replied:
@jhendersonHDF - I believe that I've fixed this, can you please re-test?

jhendersonHDF replied:
I was just paying attention to the CI checks. Since they're running again we should see soon.

fortnern commented on src/H5ES.c:
```diff
@@ -102,11 +103,62 @@ H5EScreate(void)
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5ESinsert_request(hid_t es_id, hid_t connector_id, void *request)
```
Should this include debugging parameters? Either explicitly or through a macro like the _async APIs?

qkoziol replied:
I trend in the "no" direction, since those debugging parameters are designed for application operations, and this API routine will only be invoked within a VOL connector.

fortnern commented on src/H5ESint.c:
```diff
@@ -369,6 +553,91 @@ H5ES__handle_fail(H5ES_t *es, H5ES_event_t *ev)
+    HDassert(ev);
+
+    /* Check for event that hasn't actually completed */
+    if (H5VL_REQUEST_STATUS_IN_PROGRESS == ev_status || H5VL_REQUEST_STATUS_CANT_CANCEL == ev_status)
```
Shouldn't this be an assertion?

qkoziol replied:
Make sense, done.

fortnern commented on src/H5FDdevelop.h:
```diff
@@ -0,0 +1,260 @@
+ */
+#define H5FD_MEM_FARRAY_HDR       H5FD_MEM_OHDR
+#define H5FD_MEM_FARRAY_DBLOCK    H5FD_MEM_LHEAP
+#define H5FD_MEM_FARRAY_DBLK_PAGE H5FD_MEM_LHEAP
```
Hmm maybe instead of all these ad-hoc redefinitions we should rename the categories to be decriptive of the type of objects they contain (a more general description than "ohdr" for example)

fortnern replied:
Or are these already part of the public API?

fortnern commented on src/H5Spublic.h:
```diff
@@ -23,6 +23,7 @@
 
 /* Define atomic datatypes */
 #define H5S_ALL       (hid_t)0
+#define H5S_BLOCK     (hid_t)1
```
Is H5S_BLOCK fully implemented?

qkoziol replied:
Yes, with tests, etc.

fortnern commented on src/H5VL.c:
```diff
@@ -911,3 +944,133 @@ H5VLquery_optional(hid_t obj_id, H5VL_subclass_t subcls, int opt_type, uint64_t
+ *              UTI of: "gov.anl.alcf.cache.evict".   Registering a "suspend
+ *              background threads" operation for the asynchronous VOL connector
+ *              written at NERSC at Lawrence Berkeley National Laboratory could
+ *              have a UTI of: "gov.lbnl.nersc.async.suspend_bkg_threads".
```
Seems strange to use external (to THG) organizations as examples here

fortnern commented on src/H5VLcallback.c:
```diff
@@ -5324,12 +6509,18 @@ H5VLlink_optional(void *obj, hid_t connector_id, H5VL_link_optional_t opt_type,
+
     FUNC_LEAVE_API_NOINIT(ret_value)
-} /* end H5VLlink_optional() */
+} /* end H5VLlink_optional_vararg() */
```
No H5VLlink_optional_op()?

fortnern commented on src/H5VLcallback.c:
```diff
@@ -3091,6 +3598,51 @@ H5VL_datatype_optional(const H5VL_object_t *vol_obj, H5VL_datatype_optional_t op
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5VL_datatype_optional() */
 
+/*-------------------------------------------------------------------------
```
no datatype_optional_vararg? (I see the existing one uses varargs already, but shouldn't they all be made analogous?)

qkoziol replied:
Yes, datatype_optional_vararg is here, just a little further down in the file.

jrmainzer commented on src/H5ESint.c:
```diff
@@ -50,6 +50,11 @@
 
+/* Callback context for get time estimate (gte) operations */
+typedef struct H5ES_gte_ctx_t {
+    uint64_t *time_est; /* Pointer to accumulated time estimate */
```
It would be good to mention units both here and in the timeout field of H5ES_wait_ctx_t below.

qkoziol replied:
Done

fortnern commented on src/H5VLcallback.c:
```diff
@@ -5875,6 +7158,50 @@ H5VLobject_optional(void *obj, hid_t connector_id, H5VL_object_optional_t opt_ty
+
+    FUNC_LEAVE_API_NOINIT(ret_value)
+} /* end H5VLobject_optional_vararg() */
+
```
No H5VLobject_optional_op()?

qkoziol replied:
The dynamic operations are registered on a subclass type, so it didn't seem like a good fit to add the generic form (H5VLobject_optional_op).

fortnern commented on src/H5ESint.c:
```diff
@@ -369,6 +553,91 @@ H5ES__handle_fail(H5ES_t *es, H5ES_event_t *ev)
+                op_status = H5ES_STATUS_SUCCEED;
+
+                /* Retrieve the execution time info */
+                if (H5VL_request_specific(ev->request, H5VL_REQUEST_GET_EXEC_TIME, &ev->op_info.op_exec_ts,
```
We're concerned that this code will add unnecessary overhead to every operation. It will also require all connectors to handle this specific value. Maybe we should query this capability and only call this code if it's implemented?

qkoziol replied:
So - convert it to an optional operation?

soumagne replied:
I thought we agreed in the previous discussions that time estimates were not part of that PR? I think I'm fine with optional. If not optional, I think that should just be removed from the code at this point of time and a separate PR should be made with the specific changes that add time estimates?

qkoziol committed:
Re-add macro for unsetenv on Windows (I think it accidentally was merged out)

qkoziol committed:
Strengthen sanity check from error report to assertion

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Add units to the comments for a few fields

qkoziol committed:
Merge branch 'async_vol_register_optional' of github.com:hpc-io/hdf5 into async_vol_register_optional

qkoziol commented:
I received a few emails with comments from @fortnern, but don't see them here.  Were they deleted?

soumagne commented on src/H5ES.c:
```diff
@@ -117,24 +169,69 @@ H5EScreate(void)
+
+done:
+    FUNC_LEAVE_API(ret_value)
+} /* end H5ESget_time_estimate() */
```
Same comment as the other one, why is that routine still in this PR? I thought we agreed we won't add that just yet until we have a better prototype for it?

soumagne reviewed (changes requested):
Except the time estimate routines and corresponding VOL flags that should be removed at this point of time, the rest looks good to me.

fortnern commented:
> I received a few emails with comments from @fortnern, but don't see them here. Were they deleted?

There were a few I deleted after I realized they didn't apply. It looks like you found the ones I didn't delete.

fortnern commented:
I've finished reviewing the changes. Looks good to me once the outstanding comment threads are resolved.

qkoziol committed:
Switch 'get execution time' operation for async request tokens to be an optional operation and query if connector supports operation before retrieving it.

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Remove H5ESget_time_estimate

qkoziol committed:
Merge branch 'async_vol_register_optional' of github.com:hpc-io/hdf5 into async_vol_register_optional

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Create developer header for datatype routines and move type conversion register/unregister routines there.

qkoziol committed:
Simplify internal H5VL_setup_name_args and H5VL_setup_idx_args routines

qkoziol committed:
Add H5VLlink_optional_op, allowing dynamicly registered optional operations for the link VOL subclass, also added H5VL_loc_params argument to the link 'optional' callback to allow them to work correctly.

qkoziol committed:
Run bin/format_source on current code

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Add H5VLobject_optional_op, allowing dynamicly registered optional operations for the object VOL subclass, also added H5VL_loc_params argument to the object 'optional' callback to allow them to work correctly.

qkoziol committed:
Run bin/format_source on current code

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Revert "Switch 'get execution time' operation for async request tokens to be an optional operation and query if connector supports operation before retrieving it."

This reverts commit 5ac92014da2682bdba62d7a2524b8d90e38f6b19.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

github-actions[bot] committed:
Committing clang-format changes

qkoziol committed:
Convert attribute 'get' operation to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Merge branch 'async_vol_register_optional' of github.com:hpc-io/hdf5 into async_vol_register_optional

qkoziol committed:
Update tracing macros

qkoziol committed:
Convert attribute 'specific' operation to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Convert dataset 'get' and 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Also, minor tweaks to attribute 'get' and 'specific' operation parameters.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Convert datatype 'get' and 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Also, minor tweaks to H5O_refresh_metadata arguments.

qkoziol committed:
Reduce warnings

qkoziol committed:
Reduce warnings

qkoziol committed:
Track change to datatype 'get' callback

qkoziol committed:
Fix bug with file pointer getting invalidated when object closed

qkoziol committed:
Reformat source

qkoziol committed:
Convert file and group VOL classes 'get' and 'specific' operations to use
struct-of-tagged-union pattern for VOL callback arguments, instead of using
varargs.  Also small cleanup to the attribute get name operation.  Also moved
'mount' and 'unmount' operations to be group specific operations, instead of
file specific, to better align with their behavior (mounted files are on
groups, so a group is what is operated on).

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Remove remainder of merge conflict marking

qkoziol committed:
Convert link VOL class 'create' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Remove some unused local variables

qkoziol committed:
Convert link VOL class 'get' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Also refactor 'get name by idx' routines to return actual length of name with a parameter instead of the return value, and move some callback context structs for the link interface from the private header file into the source code module, to reduce their visibility scope.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Update tracing macros

qkoziol committed:
Convert link VOL class 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Convert object VOL class 'get' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Convert object VOL class 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Also refactor H5G_loc_exists, et al, to return 'exists' flag in a parameter and errors with the function return value, instead of overloading both into the return value.  And, corrected logic error in test/links.c around non-existant objects in a file.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Convert request VOL class 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.

qkoziol committed:
Convert blob VOL class 'specific' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Also removes the H5VL_BLOB_GETSIZE operation, as it's unused in the library and the blob ID size for a container is now returned with H5VL_FILE_GET_CONT_INFO.

qkoziol committed:
Add 'const' to several parameters that are only queried.

qkoziol committed:
Convert all VOL classes' 'optional' operations to use struct-of-tagged-union pattern for VOL callback arguments, instead of using varargs.  Convert several 'get' routines to return the length of an array in a parameter instead of combining it into the return value.  Move several routines to be in less public namespace.  Correct direct_chunk test to verify that parameters aren't modified on error.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Switch get/specific/optional VOL callback argument structures to be 'async-friendly'.  Also other minor cleanups and bug-fixes.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Add H5Pset_dataset_io_hyperslab_selection / H5S_PLIST feature, to allow skipping H5Dget_space + H5Sselect_hyperslab for async operation

qkoziol committed:
Add dynamic optional operations for request objects

qkoziol committed:
Update dynamic operation test for optional request operations

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Update a comment for an operation argument

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Run trace and format_source scripts

github-actions[bot] committed:
Committing clang-format changes

qkoziol commented:
> Except the time estimate routines and corresponding VOL flags that should be removed at this point of time, the rest looks good to me.

These have been removed.

soumagne commented on src/H5Dpublic.h:
```diff
@@ -677,7 +678,7 @@ H5_DLL herr_t H5Dget_chunk_info_by_coord(hid_t dset_id, const hsize_t *offset, u
  */
-H5_DLL herr_t H5Dchunk_iter(hid_t dset_id, H5D_chunk_iter_op_t cb, void *op_data);
+H5_DLL herr_t H5Dchunk_iter(hid_t dset_id, hid_t dxpl_id, H5D_chunk_iter_op_t cb, void *op_data);
 
```
Is it not an API change? should not you do a v2?

qkoziol replied:
No, this was only merged to develop a week ago and there's no BW compatibility issues.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

soumagne commented on src/H5Spublic.h:
```diff
@@ -21,8 +21,12 @@
+#define H5S_ALL   0 /* (hid_t) */
+#define H5S_BLOCK 1 /* (hid_t) */
+#define H5S_PLIST 2 /* (hid_t) */
+
```
I feel like you should add what they mean instead of hid_t ?

qkoziol replied:
I was just extending the list (which is like that in develop).

soumagne replied:
Are they documented somewhere? I have no idea what they are

soumagne replied:
I think it would not hurt to add a one-line description of their effect / when to use them, it does not seem to be in H5Dpublic.h either.

qkoziol replied:
Yes, definitely a good thing to do:  https://github.com/HDFGroup/hdf5/issues/688

soumagne reviewed (approved):
Looks good to me, just a few minor comments.

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

qkoziol committed:
Merge remote-tracking branch 'origin/develop' into async_vol_register_optional

github-actions[bot] committed:
Committing clang-format changes
